doc/developer/platform/formalism.md

```diff
@@ -13,57 +13,370 @@ It does not intend to describe the **current behavior** of the system.
+collection version at time `t`. We can do this by applying all updates which
+occurred before or at `t` to the empty collection version. This lets us go from
+an update-set representation to a time-function representation.
+
+```
+read(updates, t) =
+  apply(cv0, { u = [data, time, diff] | u in updates and time <=t t })
+```
+
+Similarly, given a time-function representation `tvc`, we can obtain an
+update-set representation by considering each point in time and finding the
+difference between the version at that time and the product of all updates from
+all prior times:
+
+```
+updates(tvc, t) = diff(t,
+                       apply(cv0, prior-updates(tvc, t)),
+                       read(tvc, t))
+```
+
+The updates prior to `t` are simply the union of the updates for all times
+prior to `t`:
+
+```
+prior-updates(tvc, t) = ⋃ updates(tvc, prior-t) for all prior-t <t t
+```
+
+These mutually-recursive relations bottom out in the initial time `t0`, and the
+corresponding initial collection version `cv0`. Using these definitions, we
+can inductively transform an entire time-function representation `tvc` into a
+set of updates for all times by taking the union over all times:
+
+```
+updates(tvc) = ⋃ updates(tvc, t) for all t in T
+```
+
+This may seem a bit abstract---the formalism is shaped this way because
+collection versions are only partially ordered. If we take our times `T` to be
+the (totally ordered) naturals, we obtain a much simpler definition: the update
+set for a given time `t` is simply the difference between the version at `t-1`
+and the version at `t`.
+
+```
+updates(tvc, t) = diff(t, read(tvc, t - 1), read(tvc, t))
+```
+
+### Update representation of pTVCs
+
+We have argued that the time-function and update-set representations of a TVC
+are equivalent. Similarly, we can also represent a pTVC as a set of updates
+together with `since` and `upper` frontiers: `ptvc = [updates, since, upper]`.
+We use the same definition for reads as we gave for TVCs above. We wish to
+ensure that every such pTVC maintains the correctness invariant: `read(ptvc, t)
+= read(tvc, t)` so long as `t` is at least `since` and not later than `upper`.
+
+First, consider the empty pTVC `ptvc0 = [{}, {t0}, {t0}]`. `ptvc0` is trivially
+correct since there are no times not later than than the upper frontier `{t0}`.
+
+Next we consider two ways a pTVC could evolve, and wave our hands frantically
+while claiming that each evolution preserves correctness. This is not an
+exhaustive proof--we do not discuss, for example, the question of a pTVC
+which contains incorrect updates later than its upper frontier. However, this
+sketch should give a rough feeling for how and why Materialize works.
+
+### pTVC append
+
+Let `ptvc = [ptvc-updates, since, upper]` be a correct view of some TVC's
+updates `tvc_updates`. Now consider appending a window of new updates to this
+pTVC, and advancing `upper` to `upper'`: `ptvc' = append(ptvc, upper',
+new-updates)`.
+
```

Comment:
I may have missed it, in which case sorry, but it seems like this is crying out for a (notation failure) dTVC that is the differential of the TVC, which is all updates past and future, at which point a "correct pTVC in update form" has some relation to the dTVC (NB: not "the updates geq since and not geq upper" unfortunately). pTVC append should just be "add in those new updates from the dTVC", which we get to but with hand waving.

Will ponder if this could be more immediately obvious, with tweaked pre-work on the dTVC concept.

Reply:
I think if you have a dTVC, then a "correct pTVC in update form" has as updates the union of "the updates of the dTVC geq since and not geq upper" and "sufficient updates to cause the accumulations at each time in since to equal the TVC at that time". If you believed this, then it is probably clearer that pTVC append is "easy".

No opinion on whether this is helpful pedagogically, though.

Reply:
> a ... dTVC that is the differential of the TVC

This is what we defined in "Update-set representations of TVCs and pTVCs". We call this the "update-set representation" of a TVC, as opposed to the "time-function representation". I didn't want to call it dTVC because we *also* then have a dpTVC. Instead, I opted to show an isomorphism between update-set and time-function representations, which lets us just talk about "a TVC" while using whichever representation is convenient. :)

A picture might help here--I'll draw one.

> pTVC append should just be "add in those new updates from the dTVC", which we get to but with hand waving.

That's exactly what's happening here. :-)

> I think if you have a dTVC, then a "correct pTVC in update form" has as updates the union of "the updates of the dTVC geq since and not geq upper" and "sufficient updates to cause the accumulations at each time in since to equal the TVC at that time". If you believed this, then it is probably clearer that pTVC append is "easy".

That's exactly what this section does! :D

Reply:
This is probably "I came to the same conclusion by a different path" and wanted to tell you about the path I took. No reason to fix anything here.